### PR TITLE
Add comment manager and header creator

### PR DESCRIPTION
**Pull request by maiste**

### CODE REVIEW

File: src/core/comment.ml [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+(* Return a block value *)
+let get_block v =
+  (
+    try Ok (get_list get_string v)
+    with _ -> Error "[Error] Block needs an array of strings"
+  ) >>=
+  (
+    fun l ->
+      try Ok (List.nth l 0 , List.nth l 1, List.nth l 2)
```
Comment (xavitator) [inline src/core/comment.ml]: Should we fix the length of the 'block' list to 3 and return an error instead ?

File: src/core/comment.ml [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+      with _ -> Error "[Error] Block needs exactly three strings"
+  ) >>= (fun t ->  Ok (block t))
+
+(* Get ext value *)
+let get_ext v =
+  let rec aux l =
```
Comment (xavitator) [inline src/core/comment.ml]: You can use Conf.member and it will take only 3 lines.

File: src/core/comment.ml [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  match ext, ext_type with
+  | Ok e, Ok t -> Ok (e, t)
+  | Ok _, Error a -> Error a
+  | Error a, _ -> Error a
+
+
+
+(* Build a default list *)
+let build_map lst =
+  let rec from_list lst v m =
```
Comment (xavitator) [inline src/core/comment.ml]: You could use List.fold_left.

File: src/core/comment.ml [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    match lst with 
+    | [] -> m
+    | h::q ->  Printf.printf "Insert into %s\n" h ;
+        Map_Str.add h v m |> from_list q v
+  in
+  let rec aux lst m =
```
Comment (xavitator) [inline src/core/comment.ml]: Use List.fold_left too.

File: src/core/comment.ml [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+(* Update build final list *)
+let user_map js =
+  let m = build_map default in
+  match Conf.member js ["comments"] with
+  | Ok elt ->
+    Printf.printf "Ok" ;
+    (match elt with 
+      | `A lst -> 
+          build_user_map lst >>=
```
Comment (xavitator) [inline src/core/comment.ml]: add map to argument of build_user_map could be more efficient than Map.union.

File: src/core/comment.ml [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+let build_header header ext m =
+  let f s acc elt = (s^elt)::acc in
+  let map =
+    try
+      Map_Str.find ext m 
+    with Not_found -> Inline "#"
+  in
+  match map with
+  | Block (beg, mid ,ed) -> 
+      Ok (ed::(List.fold_left (f mid) [beg] header) |> List.rev)
```
Comment (xavitator) [inline src/core/comment.ml]: Split it in few lines.

File: src/core/comment.ml [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+  | Inline mid -> Ok(List.fold_left (f mid) [] header |> List.rev)
```
Comment (xavitator) [inline src/core/comment.ml]: same as above.

File: src/core/comment.ml [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+  | Error a, _ -> Error a
+
+
+
+(* Build a default list *)
+let build_map lst =
+  let rec from_list lst v m =
+    match lst with 
+    | [] -> m
+    | h::q ->  Printf.printf "Insert into %s\n" h ;
```
Comment (xavitator) [inline src/core/comment.ml]: printf log.

File: src/core/comment.ml [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+(* Build user map *)
+let build_user_map l =
+  let rec aux l acc =
+    match l with
+    | [] -> Ok (build_map (List.rev acc))
+    | h::q -> 
+        let o_list =
+          try
+            Printf.printf("Try");
```
Comment (xavitator) [inline src/core/comment.ml]: printf log.

File: src/core/comment.ml [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+  in aux l []
+
+
+
+(* Update build final list *)
+let user_map js =
+  let m = build_map default in
+  match Conf.member js ["comments"] with
+  | Ok elt ->
+    Printf.printf "Ok" ;
```
Comment (xavitator) [inline src/core/comment.ml]: printf log.

**Review by xavitator — CHANGES_REQUESTED** [review]
Remove all Printf log.

File: src/core/comment.ml (hunk 1 above)
Comment (maiste) [reply src/core/comment.ml]: Already done with this function.

File: src/core/comment.ml (hunk 2 above)
Comment (maiste) [reply src/core/comment.ml]: I can't because it use a (string, value) list

File: src/core/comment.ml (hunk 3 above)
Comment (maiste) [reply src/core/comment.ml]: Replace but let it in a function to be more readable :)

**Review by xavitator — APPROVED** [review]

**Pull request merged**